buildSrc/subprojects/buildquality/src/main/kotlin/org/gradle/gradlebuild/buildquality/incubation/IncubatingApiReportTask.kt

```diff
@@ -0,0 +1,274 @@
+        JavaParser.parse(file).run {
+            solver.inject(this)
+            findAll(Node::class.java)
+                .filter {
+                    it is NodeWithAnnotations<*> &&
+                        it.annotations.any { it.nameAsString == "Incubating" }
+                }.map {
+                    val node = it as NodeWithAnnotations<*>
+                    val version = findVersionFromJavadoc(node)
+                        ?: speculateVersion(srcDir, file, hashToVersionCache)
+                    Pair(version, nodeName(it, this, file))
+                }.forEach {
+                    versionToIncubating.getOrPut(it.first) {
+                        mutableSetOf()
+                    }.add(it.second)
+                }
+        }
+
+
+    private
+    fun findVersionFromJavadoc(node: NodeWithAnnotations<*>): String? = if (node is NodeWithJavadoc<*>) {
+        node.javadoc.map {
+            it.blockTags.find {
+                it.tagName == "since"
+            }?.content?.elements?.find {
+                it is JavadocSnippet
+            }?.toText()
+        }.orElse(null)
+    } else {
+        null
+    }
+
+    private
+    fun nodeName(it: Node?, unit: CompilationUnit, file: File) = when (it) {
+        is EnumDeclaration -> tryResolve({ it.resolve().qualifiedName }) { inferClassName(unit) }
+        is ClassOrInterfaceDeclaration -> tryResolve({ it.resolve().qualifiedName }) { inferClassName(unit) }
+        is MethodDeclaration -> tryResolve({ it.resolve().qualifiedSignature }) { inferClassName(unit) }
+        is AnnotationDeclaration -> tryResolve({ it.resolve().qualifiedName }) { inferClassName(unit) }
+        is NodeWithSimpleName<*> -> it.nameAsString
+        else -> unit.primaryTypeName.orElse(file.name)
+    }
+
+    private
+    fun inferClassName(unit: CompilationUnit) = "${unit.packageDeclaration.map { it.nameAsString }.orElse("")}.${unit.primaryTypeName.orElse("")}"
+
+    private
+    inline fun tryResolve(resolver: () -> String, or: () -> String) = try {
+        resolver()
+    } catch (e: Throwable) {
+        or()
+    }
+
+    private
+    inline fun <T> File.cmd(vararg cmd: String, action: BufferedReader.() -> T): T = ProcessBuilder(cmd.asList())
+        .directory(this)
+        .redirectOutput(ProcessBuilder.Redirect.PIPE)
+        .redirectError(ProcessBuilder.Redirect.PIPE)
+        .start()
+        .run {
+            waitFor(5, TimeUnit.SECONDS)
+            action(inputStream.bufferedReader(Charsets.UTF_8))
+        }
+
+    private
+    fun speculateVersion(srcDir: File, f: File, hashToVersionCache: MutableMap<String, String>): String {
+        val hash = srcDir.cmd("git", "log", "--format=%H", "--diff-filter=A", "--", f.toRelativeString(srcDir)) {
+            readText().trim()
+        }
+        return speculateVersionFromHash(hash, hashToVersionCache)
+    }
```

Comment:
Wow! I guess we can remove this as soon as we've retired the really old stuff, as we require `@since` on every new API.